### PR TITLE
Fix `Couldn't parse: falling back to Raw` for empty JSON array

### LINKED ISSUE #6603: `Couldn't parse: falling back to Raw` for empty JSON array
#### Problem Description
The following response packet:

```
HTTP/2.0 200 
date: Sun, 14 Jan 2024 23:30:12 GMT
content-type: application/json; charset=utf-8
content-length: 2
server: nginx/1.18.0 (Ubuntu)
cache-control: no-store
x-amzn-trace-id: [REDACTED]
x-frame-options: DENY
set-cookie: [REDACTED]
strict-transport-security: max-age=31536000

[]
```

is displayed as "Couldn't parse: falling back to Raw", but it's clearly JSON from the content type and an empty array is a normal JSON response that shouldn't trigger this warning. In fact, top level `null`, booleans, numbers and strings should be just fine as well, I don't know if they do or don't.

#### Steps to reproduce the behavior:
Set up a JSON endpoint that returns `[]` with a `content-type: application/json; charset=utf-8` header and open it in mitmproxy

#### System Information
```
Mitmproxy: 10.2.1 binary
Python:    3.12.1
OpenSSL:   OpenSSL 3.1.4 24 Oct 2023
Platform:  macOS-14.2.1-x86_64-arm-64bit
```

### PR DESCRIPTION
**Pull request by haanhvu**
#### Description

fix #6603 

#### Checklist

 - [x] I have updated tests where applicable.
 - [ ] I have added an entry to the CHANGELOG.

### CODE REVIEW

**Comment by Prinzhorn** [comment]
You're still only testing objects, in fact all the tests are only testing objects. Correct tests to add would be:

```py
assert list(json.format_json(None)) == [[("Token_Keyword_Constant", "null")]]
assert list(json.format_json(True)) == [[("Token_Keyword_Constant", "true")]]
assert list(json.format_json(1)) == [[("Token_Literal_Number", "1")]]
assert list(json.format_json("test")) == [[("Token_Literal_String", '"test"')]]
assert list(json.format_json([])) == [[("text", "[]")]]
```

The first assertion is also broken, it doesn't actually test anything:

https://github.com/mitmproxy/mitmproxy/blob/b389f63cef99c40116484d1812bdb87ce67d6e60/test/mitmproxy/contentviews/test_json.py#L19

But anyway, even `assert list(json.format_json([])) == [[("text", "[]")]]` passes. So the issue does not appear to be in `format_json` after all. I'm sure you can find out why we are ending up here:

https://github.com/mitmproxy/mitmproxy/blob/b389f63cef99c40116484d1812bdb87ce67d6e60/mitmproxy/contentviews/__init__.py#L205-L217

Something with the empty array seems to break some part of the contentview. Feel free to add the tests I mentioned above anyway for good measure.

Btw. here's how I test this:

http.txt

```
HTTP/2.0 200
content-type: application/json; charset=utf-8
content-length: 2

[]
```

```sh
ncat -vl 1337 < http.txt

# Other terminal
curl --proxy localhost:8080 http://localhost:1337
```

Now you can open the response in mitmproxy.

**Comment by haanhvu** [comment]
@Prinzhorn I added the Exception to the message [here](https://github.com/mitmproxy/mitmproxy/pull/6619/files#diff-28e8f615c63f4c0c1cdc2b9510418dafcdeff2ae67f962510b2e943fea1932d3R206).

This is what I got when tested locally: 
`list index out of range: : Couldn't parse: Falling back to Raw`

So it seems like the empty list is iterated through somewhere?

I'll debug further tomorrow. If you have any insights, pls share. Thanks!

**Comment by haanhvu** [comment]
@Prinzhorn I fixed it:
Screenshot from 2024-01-29 23-39-45

Turned out list's first item is accessed for GraphQL.

Could you review?

File: mitmproxy/contentviews/graphql.py [hunk 1]
```diff
@@ -30,6 +30,8 @@ def is_graphql_query(data):
 
 
 def is_graphql_batch_query(data):
+    if len(data) == 0:
+        return
     return isinstance(data, list) and isinstance(data[0], dict) and "query" in data[0]
```
Comment (Prinzhorn) [inline mitmproxy/contentviews/graphql.py:35]: I think you should just add the condition here

```py
return isinstance(data, list) and len(data) > 0 and isinstance(data[0], dict) and "query" in data[0]
```

**Comment by Prinzhorn** [comment]
Nice! Please remove all unrelated changes (e.g. error message and you've removed some tests). And then there should be a test case in https://github.com/mitmproxy/mitmproxy/blob/main/test/mitmproxy/contentviews/test_graphql.py

**Comment by haanhvu** [comment]
> Please remove all unrelated changes (e.g. error message and you've removed some tests)

@Prinzhorn I retrieved more details to the error message to help debugging. Do you think they should stay?

File: mitmproxy/contentviews/graphql.py (hunk 1 above)
Comment (haanhvu) [reply mitmproxy/contentviews/graphql.py:35]: If I do this then `isinstance(data[0], dict) and "query" in data[0]` will be checked and return `list index out of range`

Comment (haanhvu) [reply mitmproxy/contentviews/graphql.py:35]: Ah, sorry, my bad. I'll change it your way.

**Comment by haanhvu** [comment]
> > Please remove all unrelated changes (e.g. error message and you've removed some tests)
> 
> @Prinzhorn I retrieved more details to the error messages to help debugging. Do you think they should stay?

@Prinzhorn I adopted all your suggestions except the error messages as I asked here.

Please let me know what you think. Thanks!

File: mitmproxy/contentviews/__init__.py [hunk 2]
```diff
@@ -202,8 +202,9 @@ def get_content_view(
         desc, content = ret
         error = None
     # Third-party viewers can fail in unexpected ways...
-    except Exception:
-        desc = "Couldn't parse: falling back to Raw"
+    except Exception as e:
+        desc = str(e) + ": Couldn't parse: Falling back to Raw"
+        traceback.print_exc()
```
Comment (mhils) [inline mitmproxy/contentviews/__init__.py:207]: ´The changes here in this file are unrelated to the bugfix, please remove them.

(Without going into too much detail: The existing API contracts are weird here, and a return value of None does not only indicate no view. But that needs to be fixed properly.)

File: test/mitmproxy/contentviews/test_json.py [hunk 3]
```diff
@@ -13,6 +13,7 @@ def test_parse_json():
         b'{"foo" : "\xe4\xb8\x96\xe7\x95\x8c"}'
     )  # utf8 with chinese characters
     assert json.parse_json(b'{"foo" : "\xFF"}') is json.PARSE_ERROR
+    assert json.parse_json(b'{"foo": []}') is not json.PARSE_ERROR
```
Comment (mhils) [inline test/mitmproxy/contentviews/test_json.py:16]: This specific test method is testing for parsing errors. Is this additional test case here adding any value to that? I feel it's not, so let's remove it. The additions below are all good and cover it sufficiently!

**Review by mhils — COMMENTED** [review]
Thanks, this looks pretty good now!

**Comment by haanhvu** [comment]
@mhils I resolved all your comments

File: mitmproxy/contentviews/graphql.py [hunk 4]
```diff
@@ -30,7 +30,12 @@ def is_graphql_query(data):
 
 
 def is_graphql_batch_query(data):
-    return isinstance(data, list) and isinstance(data[0], dict) and "query" in data[0]
+    return (
+        len(data) > 0
```
Comment (Prinzhorn) [inline mitmproxy/contentviews/graphql.py:34]: You've put this in a different order than what I suggested. This will fail for anything that doesn't have a length, e.g. `null`

```
HTTP/2.0 200
content-type: application/json; charset=utf-8
content-length: 4

null
```

Comment (haanhvu) [reply mitmproxy/contentviews/graphql.py]: Yeah missed that case. I changed it back to your order.

**Review by mhils — APPROVED** [review]
Thank you two! 😃 🍰
